docs/runtime/dotnet-metrics.md

```diff
@@ -0,0 +1,679 @@
+<!--- Hugo front matter used to generate the website version of this page:
+linkTitle: .NET
+--->
+
+# Semantic Conventions for .NET (CLR) Metrics
+
+**Status**: [Experimental][DocumentStatus]
+
+This document describes semantic conventions for .NET CLR runtime metrics in OpenTelemetry.
+
+<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->
+
+<!-- toc -->
+
+- [CLR Garbage Collection](#clr-garbage-collection)
+  - [Metric: `dotnet.gc.collections.count`](#metric-dotnetgccollectionscount)
+  - [Metric: `dotnet.gc.objects.size`](#metric-dotnetgcobjectssize)
+  - [Metric: `dotnet.gc.allocations.size`](#metric-dotnetgcallocationssize)
+  - [Metric: `dotnet.gc.committed_memory.size`](#metric-dotnetgccommitted_memorysize)
+  - [Metric: `dotnet.gc.heap.size`](#metric-dotnetgcheapsize)
+  - [Metric: `dotnet.gc.heap.fragmentation.size`](#metric-dotnetgcheapfragmentationsize)
+  - [Metric: `dotnet.gc.pause.time`](#metric-dotnetgcpausetime)
+- [Just-In-Time (JIT) Compiler](#just-in-time-jit-compiler)
+  - [Metric: `dotnet.jit.il_compiled.size`](#metric-dotnetjitil_compiledsize)
+  - [Metric: `dotnet.jit.methods_compiled.count`](#metric-dotnetjitmethods_compiledcount)
+  - [Metric: `dotnet.jit.compilation.time`](#metric-dotnetjitcompilationtime)
+- [Thread pool](#thread-pool)
+  - [Metric: `dotnet.thread_pool.threads.count`](#metric-dotnetthread_poolthreadscount)
+  - [Metric: `dotnet.thread_pool.work_items.count`](#metric-dotnetthread_poolwork_itemscount)
+  - [Metric: `dotnet.thread_pool.queue.length`](#metric-dotnetthread_poolqueuelength)
+- [General](#general)
+  - [Metric: `dotnet.monitor.lock_contention.count`](#metric-dotnetmonitorlock_contentioncount)
+  - [Metric: `dotnet.timers.count`](#metric-dotnettimerscount)
+  - [Metric: `dotnet.assemblies.count`](#metric-dotnetassembliescount)
+  - [Metric: `dotnet.exceptions.count`](#metric-dotnetexceptionscount)
+
+<!-- tocstop -->
+
+## CLR Garbage Collection
+
+**Status**: [Experimental][DocumentStatus]
+
+**Description:** .NET Common Language Runtime (CLR) metrics relating to garbage collection, captured under the namespace `dotnet.gc.*`.
+
+### Metric: `dotnet.gc.collections.count`
```

Comment:
My only concern would be if there are other things related to the GC process which could/might be counted. This is currently specific about which function of the GC this metric relates to.

Reply:
>to avoid garbage collection collections

Sorry I didn't follow, where does the 'collection collections' come from?

> My only concern would be if there are other things related to the GC process which could/might be counted

One thing I could imagine being counted in the future is the number of heaps. The terminology is a bit overloaded but the GC internally sub-divides the memory it manages into areas it calls 'heaps', and then people also refer to the total set of memory the GC manages as a single 'heap'. So you have a big heap that is internally subdivided into multiple smaller heaps. One way you can observe it is via setting the [GCHeapCount GC setting](https://learn.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#ways-to-specify-the-configuration). In current versions of the runtime this number is set at process startup, but work is underway so that the number of heaps will dynamically adjust based on application workload. That might be the kind of thing someone would want to track with a metric over time.

I agree that dotnet.gc.count feels more ambiguous to me, but I want to understand what that 'collection collections' concern was first.

Reply:
the gc.collection is garbage collection collection. I agree that `gc.count` feels incomplete and limited, let's keep gc.collection.